Storage/ThirdParty/SwiftData.swift

```diff
@@ -160,15 +160,8 @@ open class SwiftData {
     func withConnection<T>(_ flags: SwiftData.Flags, synchronous: Bool = false, _ callback: @escaping (_ connection: SQLiteDBConnection) throws -> T) -> Deferred<Maybe<T>> {
         let deferred = DeferredDBOperation<Maybe<T>>()
 
-        /**
-         * We use a weak reference here instead of strongly retaining the connection because we don't want
-         * any control over when the connection deallocs. If the only owner of the connection (SwiftData)
-         * decides to dealloc it, we should respect that since the deinit method of the connection is tied
-         * to the app lifecycle. This is to prevent background disk access causing springboard crashes.
-         */
-        weak var conn = getSharedConnection()
```

Comment:
Existing behavior you've changed here:

1. Callers `A`, `B`, `C` all call `withConnection`.
2. All of them weakly bind the same connection instance.
3. `A` begins to run. It strongly holds the connection.
4. The app is backgrounded and the connection is queued to close. (I'm a bit wooly right now about how that happens.)
5. The weak references held by B and C are now `nil`.
6. B gets to the front of the queue. In the existing code, B finds that its connection is `nil` and doesn't perform its write.
7. The app is foregrounded. `closed` is now `false`.
8. C runs.

I believe the expected behavior is that C doesn't do its work: the database was closed and reopened, so its assumptions might be invalidated (e.g., it implicitly depends on B).

Existing: C's connection reference is still `nil`, so it does no work. New arrivals to the DB queue will cause the connection to be opened and be able to do their work, but C never will.

With your patch I believe that when C gets to the front of the queue, it'll find that `closed` is `false` but `sharedConnection` is `nil`, and it'll reopen the DB and perform its writes.

In order for this change to preserve behavior, you need to somehow empty the queue (or cancel all of its items) when the connection is closed.

What do you think?

Reply:
btw, if you make this change, delete `getSharedConnection`.

Reply:
One approach to this is to _switch `SwiftData` instances_ when reopening the DB.

Reply:
I believe our current closing approach is to call `Profile.shutdown`, which calls `BrowserDB.forceClose`, which calls `SwiftData.forceClose`.

That currently calls `sharedConnectionQueue.sync`. Assuming that doesn't change, it means the DB will only be closed after any existing operations have run.

If you comment the hell out of that entry point, this change should be safe — it's impossible for `B` and `C` to run with a closed connection if they were queued with an open connection. But check that!

Reply:
@rnewman checking all these things, and adding a test case.
This patch went a bit sideways at first, am closing the PR because I did a cleaner PR on a new branch.